src/normalize/index.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * @typedef {{log: {pages: Object[], entries: Object[]}}} HTTPArchive
+ * @typedef {{log: {pages: Object<{sleep?:[]|null}>[], entries: Object<{sleep?:[]|null}>[]}}} K6HTTPArchive
+ * @typedef {{startedDateTime?: string, id?: string, pageref?: string}} HARNode
+ * @typedef {{id?: string, pageRef?: string, date: Date}} TimelineRef
+ */
+
+const { SleepPlacement } = require('../enum')
+const { DEFAULT_OPTIONS } = require('../constants')
+const MIN_SLEEP = 500
+
+/**
+ * Validate archive without iterating nodes
+ * @param {HTTPArchive} archive
+ * @return {boolean}
+ */
+function isValidArchive(archive) {
+  return (
+    !!archive && !!archive.log && !!archive.log.pages && !!archive.log.entries
+  )
+}
+
+function getDraft(archive) {
+  const draft =
+    typeof archive === 'string'
+      ? JSON.parse(archive)
+      : JSON.parse(JSON.stringify(archive))
+
+  const { pages = [], entries = [], ...log } = draft.log
+  draft.log = log
+
+  return [draft, pages, entries]
+}
+
+/**
+ *
+ * @param {HARNode[]} nodes
+ * @return {TimelineRef[]}
+ */
+function getTimeline(nodes) {
+  let timeline = []
+  for (let index = 0, length = nodes.length; index < length; index += 1) {
+    const node = nodes[index]
+    const { id = '', pageref: pageRef = '', startedDateTime } = node
+    const timelineRef = { id, pageRef, date: new Date(startedDateTime) }
+
+    // Add ref to timeline and attach reference to node
+    timeline.push(timelineRef)
+    node.timelineRef = timelineRef
+
+    // Check if there is a reason to exit
+    if (Number.isNaN(timelineRef.valueOf())) {
+      // Discard getTimeline and exit loop
+      timeline = null
+      break
+    }
+  }
+
+  // Sort timeline by timingRef.date (valueOf()) unless timeline is falsy
+  if (timeline) {
+    timeline.sort((refA, refB) => {
+      return refA.date - refB.date
+    })
+  }
+
+  return timeline
+}
+
+function hasParent(timelineRef, parentRef = null) {
+  if (parentRef) {
+    return timelineRef.pageRef ? timelineRef.pageRef === parentRef.id : false
+  }
+
+  return !!timelineRef.pageRef
+}
+
+function isPage(timelineRef) {
+  return timelineRef.id && !timelineRef.pageRef
+}
+
+function withSleep(node, timeline) {
+  // Dont add sleep if sleep is truthy
+  if (!node.sleep) {
+    const precedingIndex = timeline.indexOf(node.timelineRef) - 1
```

Comment:
I've tried to make it as clear as possible with naming. I don't see how it would be brittle? Perhaps I have missed something?
I could update naming/comments if that is required?

Reply:
It is brittle because the equality is based solely on a memory address. If someone wanted to update the code and add some more information to the ref, they'd have to be aware that you mustn't never, ever change that address. They wouldn't be allowed to do, e.g.:

```
const timelineRef = { /* ... */ };
node.timelineRef = timelineRef;

/* Elsewhere in code */
if (node.duration > 0) {
    node.timelineRef = {
        ...node.timelineRef,
        endData: node.startedDate + node.duration
    }
} 
```

Reference equality is just weird and unintuitive, IMO. That's why I had to stop and think about what was actually going on.

Reply:
> Have run some testing now, both from CLI and from the Test builder in the cloud app. It works nicely! I think we can round the sleeps to 1 precision point, then its OK from me 🙄

Agree - Precision was based on `k6 convert`. I will change it to 1